packages/core/src/browser/keybinding.ts

```diff
@@ -132,60 +131,166 @@ export class KeybindingRegistry {
         for (const binding of bindings) {
-            this.registerKeybinding(binding);
+            this.doRegisterKeybinding(binding, scope);
+        }
+    }
+
+    protected doRegisterKeybinding(binding: Keybinding, scope: KeybindingScope = KeybindingScope.DEFAULT) {
+        try {
+            if (!this.isKeybindingAvailable(this.keymaps[scope], binding)) {
+                throw new Error(`"${binding.keybinding}" is in collision with something else [scope:${scope}]`);
+            }
+            this.keymaps[scope].push(binding);
+        } catch (error) {
+            this.logger.warn(`Could not register keybinding:\n  ${Keybinding.stringify(binding)}\n${error}`);
         }
     }
 
     /**
-     * Register a default keybinding to the registry.
+     * Checks for keySequence collisions in a list of Keybindings
      *
-     * @param binding
+     * @param bindings the keybinding reference list
+     * @param binding the keybinding to test collisions for
      */
-    registerKeybinding(binding: Keybinding) {
+    isKeybindingAvailable(bindings: Keybinding[], binding: Keybinding): boolean {
+        const collisions = this.getKeySequenceCollisions(bindings, KeySequence.parse(
+            this.getPlateformKeybinding(binding.keybinding))
+        ).filter(b => b.context === binding.context);
+
+        if (collisions.full.length > 0) {
+            this.logger.warn('Collided keybinding is ignored; ',
+                Keybinding.stringify(binding), ' collided with ',
+                collisions.full.map(b => Keybinding.stringify(b)).join(', '));
+            return false;
+        }
+        if (collisions.partial.length > 0) {
+            this.logger.warn('Shadowing keybinding is ignored; ',
+                Keybinding.stringify(binding), ' shadows ',
+                collisions.partial.map(b => Keybinding.stringify(b)).join(', '));
+            return false;
+        }
+        if (collisions.shadow.length > 0) {
+            this.logger.warn('Shadowed keybinding is ignored; ',
+                Keybinding.stringify(binding), ' would be shadowed by ',
+                collisions.shadow.map(b => Keybinding.stringify(b)).join(', '));
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Converts special `ctrlcmd` modifier back to `ctrl` for non-OSX users in a keybinding string.
+     * (`ctrlcmd` is mapped to the same actual key as `ctrl` under non-OSX users)
+     *
+     * @param keybinding The keybinding string to convert.
+     */
+    protected getPlateformKeybinding(keybinding: string): string {
+        return isOSX ? keybinding : keybinding.replace(/\bctrlcmd\b/, 'ctrl');
+    }
+
+    /**
+     * Finds collisions for a binding inside a list of bindings (error-free)
+     *
+     * @param bindings the reference bindings
+     * @param binding the binding to match
+     */
+    protected getKeybindingCollisions(bindings: Keybinding[], binding: Keybinding): KeybindingRegistry.KeybindingsResult {
```

Comment:
Does it really need to be part of the `KeybindingRegistry` namespace? IMO that's just more difficult to read, and AFAIK it's only used in the keybinding file, if it's not really used outside, just don't export it.

Reply:
I was advised to do so in the previous review, it was as you described before.